pkg/handler/knowledgebase.go

```diff
@@ -3,29 +3,193 @@ package handler
+	if err != nil {
+		errorRes.ErrorMsg = fmt.Sprintf("failed to get user id from header: %v", err)
+		return errorRes, nil
+	}
+	// check name if it is empty
+	if req.Name == "" {
+		errorRes.ErrorMsg = "name is required"
+		return errorRes, nil
+	}
+	nameOk := isValidName(req.Name)
+	if !nameOk {
+		errorRes.ErrorMsg = "name is invalid: " + req.Name
+		return errorRes, nil
+	}
+	res, err := ph.service.Repository.CreateKnowledgeBase(ctx,
+		repository.KnowledgeBase{
+			Name:        req.Name,
+			KbID:        toIDStyle(req.Name),
+			Description: req.Description,
+			Tags:        req.Tags,
+			Owner:       uid,
+		},
+	)
+	if err != nil {
+		errorRes.ErrorMsg = fmt.Sprintf("failed to create knowledge base: %v", err)
+		return errorRes, nil
+	}
+	return &artifactpb.CreateKnowledgeBaseResponse{
+		Body: &artifactpb.KnowledgeBase{
+			Name:        res.Name,
+			Id:          res.KbID,
+			Description: res.Description,
+			Tags:        res.Tags,
+			OwnerName:   res.Owner,
+			CreateTime:  res.CreateTime.String(),
+			UpdateTime:  res.UpdateTime.String(),
+		}, ErrorMsg: "", StatusCode: 0,
+	}, nil
 }
-func (ph *PublicHandler) GetKnowledgeBases(ctx context.Context, req *artifactpb.GetKnowledgeBasesRequest) (*artifactpb.GetKnowledgeBasesResponse, error) {
-	fmt.Println("GetKnowledgeBases")
-	return nil, nil
+func (ph *PublicHandler) ListKnowledgeBases(ctx context.Context, _ *artifactpb.ListKnowledgeBasesRequest) (*artifactpb.ListKnowledgeBasesResponse, error) {
+	errorRes := &artifactpb.ListKnowledgeBasesResponse{
+		StatusCode: ErrorGetKnowledgeBasesCode,
+	}
+	// get user id from context
+	uid, err := getUserIDFromContext(ctx)
+	if err != nil {
+		errorRes.ErrorMsg = fmt.Sprintf("failed to get user id from header: %v", err)
+		return errorRes, nil
+	}
+	res, err := ph.service.Repository.ListKnowledgeBases(ctx, uid)
+	if err != nil {
+		errorRes.ErrorMsg = fmt.Sprintf("failed to get knowledge bases: %v", err)
+		return errorRes, nil
+	}
+
+	kbs := make([]*artifactpb.KnowledgeBase, len(res))
+	for i, kb := range res {
+		kbs[i] = &artifactpb.KnowledgeBase{
+			Name:        kb.Name,
+			Id:          kb.KbID,
+			Description: kb.Description,
+			Tags:        kb.Tags,
+			CreateTime:  kb.CreateTime.String(),
+			UpdateTime:  kb.UpdateTime.String(),
+			OwnerName:   kb.Owner,
+		}
+	}
+	return &artifactpb.ListKnowledgeBasesResponse{
+		Body: &artifactpb.KnowledgeBasesList{
+			KnowledgeBases: kbs,
+		},
+		ErrorMsg: "", StatusCode: 0,
+	}, nil
 }
 func (ph *PublicHandler) UpdateKnowledgeBase(ctx context.Context, req *artifactpb.UpdateKnowledgeBaseRequest) (*artifactpb.UpdateKnowledgeBaseResponse, error) {
-	fmt.Println("UpdateKnowledgeBase")
-	return nil, nil
+	errorRes := &artifactpb.UpdateKnowledgeBaseResponse{
+		StatusCode: ErrorUpdateKnowledgeBaseCode,
+	}
+	uid, err := getUserIDFromContext(ctx)
+	if err != nil {
+		errorRes.ErrorMsg = fmt.Sprintf("failed to get user id from header: %v", err)
+		return errorRes, nil
```

Comment:
@Yougigun Does it mean that even if there is an error, the backend still returns HTTP OK?

Reply:
yes. I originally used custom defined status code. now I just have returned the error to handler so that framwwork do the native error message and code.
it looks like 
```
{
    "code": 2,
    "message": "failed to update knowledge base: record not found",
    "details": []
}
```
with `http status code 500`

Reply:
Got it. Regarding the HTTP status code, you can refer to other backends. We can map the different errors to different HTTP codes.

Reply:
got it. 
In terms of HTTP status code, yes I just noticed it. does it block this pr, could I refactor it using grpc code mapping HTTP in another pr ?
Thank you for reply.